src/main/java/com/cloudbees/jenkins/support/SupportPlugin.java

```diff
@@ -260,136 +264,158 @@ public static void setLogLevel(Level level) {
-                logger.log(logRecord);
-                errorWriter.println(
-                        MessageFormat.format("Could not get content from ''{0}'' for support bundle", cDisplayName));
-                errorWriter.println("-----------------------------------------------------------------------");
-                errorWriter.println();
-                SupportLogFormatter.printStackTrace(e, errorWriter);
-                errorWriter.println();
-
-            }
-        }
-        toProcess.add(new StringContent("manifest.md", manifest.toString()));
+        appendManifestHeader(manifest);
+        List<Content> contents = appendManifestContents(manifest, errorWriter, components);
+        contents.add(new StringContent("manifest.md", manifest.toString()));
         try {
-            ZipOutputStream zip = new ZipOutputStream(outputStream);
-            try {
-                BufferedOutputStream bos = new BufferedOutputStream(zip, 16384) {
-                    @Override
-                    public void close() throws IOException {
-                        // don't let any of the contents accidentally close the zip stream
-                        super.flush();
-                    }
-                };
-                while (!toProcess.isEmpty()) {
-                    Content c = toProcess.poll();
-                    if (c == null) {
+            try (BulkChange change = new BulkChange(ContentMappings.get());
+                 ZipArchiveOutputStream binaryOut = new ZipArchiveOutputStream(new BufferedOutputStream(outputStream, 16384))) {
+                Optional<ContentFilter> maybeFilter = getContentFilter();
+                Optional<FilteredOutputStream> maybeFilteredOut = maybeFilter.map(filter -> new FilteredOutputStream(binaryOut, filter));
+                OutputStream textOut = maybeFilteredOut.map(OutputStream.class::cast).orElse(binaryOut);
+                for (Content content : contents) {
+                    if (content == null) {
                         continue;
                     }
-                    final String name = c.getName();
+                    final String name = maybeFilter.map(filter -> filter.filter(content.getName())).orElseGet(content::getName);
+                    final ZipArchiveEntry entry = new ZipArchiveEntry(name);
+                    entry.setTime(content.getTime());
                     try {
-                        ZipEntry entry = new ZipEntry(name);
-                        entry.setTime(c.getTime());
-                        zip.putNextEntry(entry);
-                        c.writeTo(bos);
+                        binaryOut.putArchiveEntry(entry);
+                        binaryOut.flush();
+                        OutputStreamSelector selector = new OutputStreamSelector(() -> binaryOut, () -> textOut);
+                        IgnoreCloseOutputStream out = new IgnoreCloseOutputStream(selector);
+                        content.writeTo(out);
+                        out.flush();
```

Comment:
What if there was a `Content` with fewer bytes than the selector uses to make its selection? Do we need an `OutputStreamSelector#forceSelection` method that is called before `out.flush`?

Reply:
Right now it looks like in `OutputStreamSelector#probeContents` that we would delay the selection (`head.hasRemaining()` is true) and not write anything unless `#close` was called.

Reply:
Excellent point. I'll update this.

Reply:
To be fair, seems unlikely in practice except for maybe `FileContent` with a tiny file, but could be bad if we later changed `DEFAULT_PROBE_SIZE`.

Reply:
I think this could be addressed by adding a call to `OutputStreamSelector::chooseStream` inside `flush`.